src/Control/Middleware/CanonicalURLMiddleware.php

```diff
@@ -0,0 +1,347 @@
+    }
+
+    /**
+     * @param bool $forceSSL
+     * @return $this
+     */
+    public function setForceSSL($forceSSL)
+    {
+        $this->forceSSL = $forceSSL;
+        return $this;
+    }
+
+    /**
+     * Generate response for the given request
+     *
+     * @param HTTPRequest $request
+     * @param callable $delegate
+     * @return HTTPResponse
+     */
+    public function process(HTTPRequest $request, callable $delegate)
+    {
+        // Handle any redirects
+        $redirect = $this->getRedirect($request);
+        if ($redirect) {
+            return $redirect;
+        }
+
+        return $delegate($request);
+    }
+
+    /**
+     * Given request object determine if we should redirect.
+     *
+     * @param HTTPRequest $request Pre-validated request object
+     * @return HTTPResponse|null If a redirect is needed return the response
+     */
+    protected function getRedirect(HTTPRequest $request)
+    {
+        // Check global disable
+        if (!$this->isEnabled()) {
+            return null;
+        }
+
+        // Check ssl
+        $forceSSL = $this->getRedirectForceSSL($request);
+        if ($forceSSL) {
+            return $forceSSL;
+        }
+
+        // Check www
+        $forceWWW = $this->getRedirectForceWWW($request);
+        if ($forceWWW) {
+            return $forceWWW;
+        }
+
+        return null;
+    }
+
+    /**
+     * Handles redirection to canonical urls outside of the main middleware chain
+     * using HTTPResponseException.
+     * Will not do anything if a current HTTPRequest isn't available
+     *
+     * @param HTTPRequest|null $request Allow HTTPRequest to be used for the base comparison
+     * @throws HTTPResponse_Exception
+     */
+    public function throwRedirectIfNeeded(HTTPRequest $request = null)
+    {
+        $request = $this->getOrValidateRequest($request);
+        if (!$request) {
+            return;
+        }
+        $response = $this->getRedirect($request);
+        if ($response) {
+            throw new HTTPResponse_Exception($response);
+        }
+    }
+
+    /**
+     * Return a valid request, if one is available, or null if none is available
+     *
+     * @param HTTPRequest $request
+     * @return mixed|null
+     */
+    protected function getOrValidateRequest(HTTPRequest $request = null)
+    {
+        if ($request instanceof HTTPRequest) {
+            return $request;
+        }
+        if (Injector::inst()->has(HTTPRequest::class)) {
+            return Injector::inst()->get(HTTPRequest::class);
+        }
+        return null;
+    }
+
+    /**
+     * Check if a redirect for SSL is necessary
+     * @param HTTPRequest $request
+     * @return HTTPResponse|null
+     */
+    protected function getRedirectForceSSL(HTTPRequest $request)
+    {
+        // Check if force SSL is enabled
+        if (!$this->getForceSSL()) {
+            return null;
+        }
+
+        // Already on SSL
+        if (Director::is_https($request)) {
```

Comment:
As reported in #7492 this will provide a false positive if the `SS_BASE_URL` is set to an https scheme even if the request isn't. I recommend checking against the request only.

Reply:
It's nothing to do with SS_BASE_URL; `Director::is_https()` prefers the Request object before checking SS_BASE_URL.

When this line is called in CanonicalURLMiddleware, the $request will always be populated (since this middleware only activates when a request is available).

Reply:
I think in that case, it shouldn't be functionally incorrect to do as you suggest, and just inline the small part of Director::is_https(). I think I see your point. :)

Reply:
Ok, this identified a painful flaw in Director::makeRelative(). For sites with multiple domains this method wouldn't work, meaning that Director::mockRequest() was failing in tests.

After a bit of work I've fixed and tested this, however. Now we can directly use the request instead of needing Director for getting request components.